File: core/src/net/natruid/jungle/systems/render/RectRenderSystem.java
package net.natruid.jungle.systems.render;

import com.artemis.Aspect;
import com.artemis.ComponentMapper;
import com.badlogic.gdx.graphics.glutils.ShapeRenderer;
import com.badlogic.gdx.math.Vector3;
import net.natruid.jungle.components.render.AngleComponent;
import net.natruid.jungle.components.render.PivotComponent;
import net.natruid.jungle.components.render.RectComponent;
import net.natruid.jungle.components.render.ScaleComponent;
import net.natruid.jungle.utils.types.RendererType;

public class RectRenderSystem extends RenderSystem {
    private ComponentMapper<RectComponent> mRect;
    private ComponentMapper<PivotComponent> mPivot;
    private ComponentMapper<ScaleComponent> mScale;
    private ComponentMapper<AngleComponent> mAngle;

    public RectRenderSystem() {
        super(Aspect.all(RectComponent.class), RendererType.SHAPE_FILLED);
    }

    @Override
    public void render(int entityId) {
        ShapeRenderer shapeRenderer = renderer.shapeRenderer;
        RectComponent rect = mRect.get(entityId);
        PivotComponent pivot = mPivot.getSafe(entityId, PivotComponent.DEFAULT);
        ScaleComponent scale = mScale.getSafe(entityId, ScaleComponent.DEFAULT);
        AngleComponent angle = mAngle.getSafe(entityId, AngleComponent.NONE);

        float originX = rect.width * pivot.xy.x;
        float originY = rect.height * pivot.xy.y;

        shapeRenderer.setColor(rect.color);
        Vector3 pos = getPos(-originX, -originY);
        shapeRenderer.rect(
            pos.x,
            pos.y,
            originX,
            originY,
            rect.width,
            rect.height,
            scale.xy.x,
            scale.xy.y,
            angle.rotation
        );
    }
}

File: core/src/net/natruid/jungle/systems/render/RectRenderSystem.kt
package net.natruid.jungle.systems.render

import com.artemis.Aspect
import com.artemis.ComponentMapper
import net.natruid.jungle.components.render.AngleComponent
import net.natruid.jungle.components.render.PivotComponent
import net.natruid.jungle.components.render.RectComponent
import net.natruid.jungle.components.render.ScaleComponent
import net.natruid.jungle.utils.types.RendererType

class RectRenderSystem : RenderSystem(
    Aspect.all(RectComponent::class.java),
    RendererType.SHAPE_FILLED
) {
    private lateinit var mRect: ComponentMapper<RectComponent>
    private lateinit var mPivot: ComponentMapper<PivotComponent>
    private lateinit var mScale: ComponentMapper<ScaleComponent>
    private lateinit var mAngle: ComponentMapper<AngleComponent>

    override fun render(entityId: Int) {
        val shapeRenderer = renderer.shapeRenderer
        val rect = mRect[entityId]
        val pivot = mPivot.getSafe(entityId, PivotComponent.DEFAULT)
        val scale = mScale.getSafe(entityId, ScaleComponent.DEFAULT)
        val angle = mAngle.getSafe(entityId, AngleComponent.NONE)

        val originX = rect.width * pivot.xy.x
        val originY = rect.height * pivot.xy.y

        shapeRenderer.color = rect.color
        val pos = getPos(-originX, -originY)
        shapeRenderer.rect(
            pos.x,
            pos.y,
            originX,
            originY,
            rect.width,
            rect.height,
            scale.xy.x,
            scale.xy.y,
            angle.rotation
        )
    }
}
